### PR TITLE
#456 Elasticsearch doc manager should accept multiple hosts

### PR DESCRIPTION
**Pull request by ghaughian**
Added support for multiple Elastic hosts to be supplied. Thus, ensuring HA of mongo-connector with regards to Elastic clusters

### CODE REVIEW

File: tests/test_elastic2.py [hunk 1]
```diff
@@ -277,5 +280,47 @@ def test_bad_int_value(self):
...
 
+class TestElasticCluster(TestElastic):
+    """Integration tests for mongo-connector + Elasticsearch Cluster."""
+
+    @classmethod
+    def setUpClass(cls):
+        """Start the cluster."""
+        super(TestElasticCluster, cls).setUpClass(_hosts=elastic_nodes)
+        cls.repl_set = ReplicaSet().start()
+        cls.conn = cls.repl_set.client()
```
Comment (ShaneHarvey) [inline tests/test_elastic2.py]: `TestElastic.setUpClass` already creates the replica set and connection so you can remove the two lines above.

File: tests/test_elastic2.py [hunk 2]
```diff
@@ -277,5 +280,47 @@ def test_bad_int_value(self):
...
+
```
Comment (ShaneHarvey) [inline tests/test_elastic2.py]: You can remove all of the methods below. Since this class is a subclass of `TestElastic` Python automatically routes the method calls to the parent class similar to how you have written it.

**Comment by ShaneHarvey** [comment]
The changes look good. The only thing I'm concerned about is how to test TestElasticCluster. The travis file launches a single Elasticsearch instance so TestElasticCluster isn't actually running any tests against a cluster. @llvtt any thoughts?

**Comment by llvtt** [comment]
It doesn't look like Travis supports running a full-blown Elastic cluster. It sets up Elastic in the "default" configuration, which is all we get.

We could spin up multiple Elastic nodes and run some tests connected only to a single node and others to all nodes, but that's probably more complicated than it's worth.

Alternatively, we could simply test that the Elastic Doc Manager can be created connected to multiple hosts (maybe assert something about `doc_manager.elastic.transport.hosts`), and test only against a single node. We assume that the Elastic client handles connections to multiple nodes transparently.

File: mongo_connector/doc_managers/elastic2_doc_manager.py [hunk 3]
```diff
@@ -82,8 +82,14 @@ def __init__(self, url, auto_commit_interval=DEFAULT_COMMIT_INTERVAL,
             client_options['use_ssl'] = True
             client_options['verify_certs'] = True
             client_options['connection_class'] = es_connection.RequestsHttpConnection
+        if type(url) is list:
+            client_options.setdefault('sniff_on_start', True)
+            client_options.setdefault('sniff_on_connection_fail', True)
+            client_options.setdefault('sniffer_timeout', 60)
```
Comment (llvtt) [inline mongo_connector/doc_managers/elastic2_doc_manager.py]: Where are these options coming from? Why set them, instead of leaving them as the default?

File: tests/test_elastic2.py [hunk 4]
```diff
@@ -32,16 +32,19 @@
...
 
     @classmethod
-    def setUpClass(cls):
-        cls.elastic_conn = Elasticsearch(hosts=[elastic_pair])
-        cls.elastic_doc = DocManager(elastic_pair,
+    def setUpClass(cls, _hosts=elastic_pair):
+        if type(_hosts) is list:
+            cls.elastic_conn = Elasticsearch(hosts=_hosts)
+        else:
+            cls.elastic_conn = Elasticsearch(hosts=[_hosts])
```
Comment (llvtt) [inline tests/test_elastic2.py]: Perhaps always pass a list to _hosts. That would simplify this.

File: tests/test_elastic2.py [hunk 5]
```diff
@@ -86,9 +89,9 @@ class TestElastic(ElasticsearchTestCase):
     """Integration tests for mongo-connector + Elasticsearch."""
 
     @classmethod
-    def setUpClass(cls):
+    def setUpClass(cls, _hosts=elastic_pair):
```
Comment (llvtt) [inline tests/test_elastic2.py]: Is this _hosts parameter being set somewhere, or is it always elastic_pair?

Comment (ghaughian) [inline tests/test_elastic2.py]: elastic_pair and elastic_nodes are variables pulled in from test/__init__.py. elastic_pair is the original variable which is a host:port pair of a single elastic node. We therefore want this to be the default behaviour should a subclass instantiate this object without explicitly specifying a list or single elastic node(s) to connect to.

File: tests/test_elastic2.py (hunk 4 above)
Comment (ghaughian) [inline tests/test_elastic2.py]: Agreed! I will update

File: mongo_connector/doc_managers/elastic2_doc_manager.py (hunk 3 above)
Comment (ghaughian) [inline mongo_connector/doc_managers/elastic2_doc_manager.py]: These are options that support connecting to more than 1 elastic node in a cluster. By passing a seed list of hosts the Elasticsearch object is able to "sniff" for other nodes in the cluster. These options enable controlling that "sniffing" behaviour.

**Comment by ShaneHarvey** [comment]
If it's not necessary to test against 2 elastic nodes we can replace your changes with a test that only checks that we can pass multiple hosts to elastic. Something akin to:

```python
TestElasticMultipleHosts(unittest.TestCase):
    def test_multiple_hosts(self):
        elastic_doc = DocManager(elastic_nodes)
        self.assertEqual(len(elastic_doc.transport.hosts), 2)
```

**Comment by llvtt** [comment]
I like that approach, @ShaneHarvey.

File: mongo_connector/doc_managers/elastic2_doc_manager.py (hunk 3 above)
Comment (ShaneHarvey) [inline mongo_connector/doc_managers/elastic2_doc_manager.py]: The docs recommend enabling this options for long running programs: http://elasticsearch-py.readthedocs.io/en/master/index.html#sniffing

Comment (ghaughian) [inline mongo_connector/doc_managers/elastic2_doc_manager.py]: Yes, we probably should. It means even if only a single host:port is provided it will be able to operate as if it is connected to an entire cluster w/failover & load balancing.

File: tests/test_elastic2.py [hunk 6]
```diff
@@ -277,5 +277,17 @@ def test_bad_int_value(self):
...
+class TestElasticMultipleHosts(TestElastic):
+    """Integration tests for mongo-connector + Elasticsearch Cluster."""
+
+    @classmethod
+    def setUpClass(cls):
+        """Start the cluster."""
+        super(TestElasticMultipleHosts, cls).setUpClass(_hosts=elastic_nodes)
+
+    def test_multiple_hosts(self):
+        self.assertEqual(len(self.elastic_doc.elastic.transport.hosts), 2)
```
Comment (ShaneHarvey) [inline tests/test_elastic2.py]: change this to:
```python
elastic_doc =  DocManager(elastic_nodes)
self.assertEqual(len(elastic_doc.elastic.transport.hosts), 2)
```

File: tests/test_elastic2.py [hunk 7]
```diff
@@ -277,5 +277,17 @@ def test_bad_int_value(self):
             self.assertNotIn('nan', doc)
             self.assertTrue(doc['still_exists'])
 
+class TestElasticMultipleHosts(TestElastic):
```
Comment (ShaneHarvey) [inline tests/test_elastic2.py]: Make this a sublcass of unittest.TestCase instead of TestElastic.

File: tests/test_elastic2.py [hunk 8]
```diff
@@ -277,5 +277,17 @@ def test_bad_int_value(self):
...
+    """Integration tests for mongo-connector + Elasticsearch Cluster."""
+
+    @classmethod
+    def setUpClass(cls):
+        """Start the cluster."""
+        super(TestElasticMultipleHosts, cls).setUpClass(_hosts=elastic_nodes)
```
Comment (ShaneHarvey) [inline tests/test_elastic2.py]: Remove this method completely.

File: tests/test_elastic2.py [hunk 9]
```diff
@@ -86,9 +86,9 @@ class TestElastic(ElasticsearchTestCase):
     """Integration tests for mongo-connector + Elasticsearch."""
 
     @classmethod
-    def setUpClass(cls):
+    def setUpClass(cls, _hosts=[elastic_pair]):
         """Start the cluster."""
-        super(TestElastic, cls).setUpClass()
+        super(TestElastic, cls).setUpClass(_hosts=_hosts)
```
Comment (ShaneHarvey) [inline tests/test_elastic2.py]: Revert these changes for the same reason.

File: tests/test_elastic2.py [hunk 10]
```diff
@@ -32,16 +32,16 @@
...
 class ElasticsearchTestCase(unittest.TestCase):
     """Base class for all ES TestCases."""
 
     @classmethod
-    def setUpClass(cls):
-        cls.elastic_conn = Elasticsearch(hosts=[elastic_pair])
-        cls.elastic_doc = DocManager(elastic_pair,
+    def setUpClass(cls, _hosts=[elastic_pair]):
+        cls.elastic_conn = Elasticsearch(hosts=_hosts)
+        cls.elastic_doc = DocManager(_hosts,
```
Comment (ShaneHarvey) [inline tests/test_elastic2.py]: Revert these changes since we are not testing everything against a multi-host ElasticDocManager.

File: tests/test_elastic2.py [hunk 11]
```diff
@@ -86,7 +86,7 @@ class TestElastic(ElasticsearchTestCase):
     """Integration tests for mongo-connector + Elasticsearch."""
 
     @classmethod
-    def setUpClass(cls):
+    def setUpClass(cls, _hosts=[elastic_pair]):
```
Comment (ShaneHarvey) [inline tests/test_elastic2.py]: Just remove this one last part and then it looks good to me.

**Comment by ShaneHarvey** [comment]
I'm going to merge this in. The tests are failing because of #21. We should document the new `sniff_*` connection defaults on the next release.
